massa-event-cache/src/controller.rs

```diff
@@ -0,0 +1,127 @@
+pub trait EventCacheController: Send + Sync {
+    fn save_events(&self, events: VecDeque<SCOutputEvent>);
+
+    fn get_filtered_sc_output_events(&self, filter: &EventFilter) -> Vec<SCOutputEvent>;
+}
+
+#[derive(Clone)]
+/// implementation of the event cache controller
+pub struct EventCacheControllerImpl {
+    /// input data to process in the VM loop
+    /// with a wake-up condition variable that needs to be triggered when the data changes
+    pub(crate) input_data: Arc<(Condvar, Mutex<EventCacheWriterInputData>)>,
+    /// Event cache
+    pub(crate) cache: Arc<RwLock<EventCache>>,
+}
+
+impl EventCacheController for EventCacheControllerImpl {
+    fn save_events(&self, events: VecDeque<SCOutputEvent>) {
+        // lock input data
+        let mut input_data = self.input_data.1.lock();
+        input_data.events.extend(events);
+        // wake up VM loop
+        self.input_data.0.notify_one();
+    }
+
+    fn get_filtered_sc_output_events(&self, filter: &EventFilter) -> Vec<SCOutputEvent> {
+        let lock_0 = self.input_data.1.lock();
+        #[allow(clippy::unnecessary_filter_map)]
+        let it = lock_0.events.iter().filter_map(|event| {
+            if let Some(start) = filter.start {
+                if event.context.slot < start {
+                    return None;
+                }
+            }
+            if let Some(end) = filter.end {
+                if event.context.slot >= end {
+                    return None;
+                }
+            }
+            if let Some(is_final) = filter.is_final {
+                if event.context.is_final != is_final {
+                    return None;
+                }
+            }
+            if let Some(is_error) = filter.is_error {
+                if event.context.is_error != is_error {
+                    return None;
+                }
+            }
+            match (
+                filter.original_caller_address,
+                event.context.call_stack.front(),
+            ) {
+                (Some(addr1), Some(addr2)) if addr1 != *addr2 => return None,
+                (Some(_), None) => return None,
+                _ => (),
+            }
+            match (filter.emitter_address, event.context.call_stack.back()) {
+                (Some(addr1), Some(addr2)) if addr1 != *addr2 => return None,
+                (Some(_), None) => return None,
+                _ => (),
+            }
+            match (
+                filter.original_operation_id,
+                event.context.origin_operation_id,
+            ) {
+                (Some(addr1), Some(addr2)) if addr1 != addr2 => return None,
+                (Some(_), None) => return None,
+                _ => (),
+            }
+            Some(event)
+        });
+
+        let mut res_0: BTreeSet<SCOutputEvent> = it.cloned().collect();
+        // Drop the lock on the queue as soon as possible to avoid deadlocks
+        drop(lock_0);
```

Comment:
it is better practice to use a scope

Reply:
I think a drop is more explicit for a reader to check if we really drop the lock as early as possible. Anyway I've added the scope + keep the drop.

Reply:
ok